components/list/src/lib.rs

```diff
@@ -57,6 +69,7 @@ pub use list_formatter::*;
 /// [CLDR spec](https://unicode.org/reports/tr35/tr35-general.html#ListPatterns)
 /// for an explanation of the different styles.
 #[derive(Copy, Clone, PartialEq, Debug)]
+#[non_exhaustive]
```

Comment:
I think we should make this exhaustive. This is tightly coupled to `ListFormatterPatternsV1`, if we ever update to a `ListFormatterPatternV2`, which would be required to add an additional style, we have to update the public API and can replace this enum as well.

Reply:
works for me!

Reply:
Well. It's an _input_, does it matter if it's actually exhaustive? People won't be matching on it either way

In general I've tended to make inputs non_exhaustive, but i'm okay with diverging here if you really feel it's stable

Note that ListFormatterPatternV2 can be added without a breaking change, and this being exhaustive would make it a breaking change to do so.

Reply:
It's not matching that is a problem, but we cannot add variants to this and keep the provider struct stable and avoid panicking/unnecessary runtime errors. The enum is used to index the array in the data struct, so if we ever add a variant the indexing code will become panicky, which cannot happen, and it doesn't make sense to make this return `Result`s just so we don't have to use a new enum in a future constructor that takes `P: ResourceProvider<V2>`

Reply:
Here's the design doc for data versioning:

https://docs.google.com/document/d/1yg_2l5FFo0aAuNi4jpgcIhIYjHqJyUoJWtMduyQ0vR8/edit

The main concern is new code with old data. The *code* knows how many variants of `ListStyle` there are, and that number is fixed at compile time. As long as Rust code has a `ListStyle` object, it is safely in range.

Therefore, new code will be able to know and handle cases when old data does not have the correct styles available. In other words, if a new ListStyle is used but only an old data struct is available, this would result in an Err result on ListFormat::try_new.

Reply:
Right, we just have to ensure that when new code is written, this check is done. I think right now we need to `allow` line 43 because the compiler cannot figure out that `style as usize` < 3, so marking the enum as non-exhaustive means adding a variant will make this code implicitly panicky.

Reply:
It won't: non_exhaustive is purely applicable for code outside the crate and doesn't affect in-crate semantics at all. I think this is the crux of the disagreement: you're talking about stuff within the crate, which this doesn't affect

Reply:
We can stick a comment there so people know to update stuff